mempool/reactor_test.go

```diff
@@ -685,3 +685,162 @@ func TestMempoolVectors(t *testing.T) {
+	secondNodeFromThird := reactors[2].Switch.Peers().Get(secondNodeID)
+
+	thirdNodeID := reactors[2].Switch.NodeInfo().ID()
+	thirdNodeFromFirst := reactors[0].Switch.Peers().Get(thirdNodeID)
+	thirdNodeFromSecond := reactors[1].Switch.Peers().Get(thirdNodeID)
+
+	firstNodeID := reactors[0].Switch.NodeInfo().ID()
+	firstNodeFromThird := reactors[2].Switch.Peers().Get(firstNodeID)
+
+	txUnique := make(map[string][]byte, len(txs))
+	for _, tx := range txs {
+		txUnique[string(tx.Hash())] = tx
+	}
+
+	// Add transactions to node 2 from node 3
+	for _, tx := range txUnique {
+		_, err := reactors[1].TryAddTx(tx, thirdNodeFromSecond)
+		require.NoError(t, err)
+	}
+
+	// Add transactions to node 3 from node 2
+	// node3.senders[tx] = node2
+	for _, tx := range txUnique {
+		_, err := reactors[2].TryAddTx(tx, secondNodeFromThird)
+		require.NoError(t, err)
+	}
+
+	// Add transactions to node 1 from node 2
+	for _, tx := range txUnique {
+		_, err := reactors[0].TryAddTx(tx, secondNode)
+		require.NoError(t, err)
+	}
+
+	// Trying to add the same transactions node 1 has received
+	// from node 2, but this time from node 3
+	// Node 1 should now ask node 3 to disable the route between
+	// a node that has sent this tx to node 3(node 2) and node1
+	for _, tx := range txUnique {
+		_, err := reactors[0].TryAddTx(tx, thirdNodeFromFirst)
+		// The transaction is in cache, hence the Error
+		require.ErrorIs(t, err, ErrTxInCache)
+	}
+
+	// Wait for the redundancy adjustment to kick in
+	time.Sleep(config.Mempool.DOGAdjustInterval)
+
+	// Make sure that Node 3 has at least one disabled route
+	require.Greater(t, len(reactors[2].router.disabledRoutes), 0)
+
+	require.True(t, reactors[2].router.isRouteDisabled(secondNodeFromThird.ID(), firstNodeFromThird.ID()))
+	txs = NewRandomTxs(numTxs, 20)
+	for _, tx := range txs {
+		_, err := reactors[0].TryAddTx(tx, thirdNodeFromFirst)
+		require.NoError(t, err)
+	}
+	reactors[0].Switch.StopPeerGracefully(secondNode)
+	// Wait for the redundancy adjustment to kick in
+	time.Sleep(config.Mempool.DOGAdjustInterval)
+	require.False(t, reactors[2].router.isRouteDisabled(secondNodeFromThird.ID(), firstNodeFromThird.ID()))
+
+	// TEST Reset on peer disconnect
+}
+
+// When a peer disconnects we want to remove all disabled route info
+// for that peer only.
+func TestDOGRemoveDisabledRoutesOnDisconnect(t *testing.T) {
+	config := cfg.TestConfig()
+	config.Mempool.DOGProtocolEnabled = true
+
+	// Put the interval to a higher value to make sure the values don't get reset
+	config.Mempool.DOGAdjustInterval = 5 * time.Second
+	reactors, _ := makeAndConnectReactors(config, 3, nil)
+
+	firstNodeID := reactors[0].Switch.NodeInfo().ID()
+
+	secondNodeID := reactors[1].Switch.NodeInfo().ID()
+	secondNode := reactors[0].Switch.Peers().Get(secondNodeID)
+
+	thirdNodeID := reactors[2].Switch.NodeInfo().ID()
+
+	reactors[0].router.disableRoute(secondNodeID, firstNodeID)
+	reactors[0].router.disableRoute(thirdNodeID, firstNodeID)
```

Comment:
In a route, source and target are peers of a node. Here a node disables a route to itself (reactors[0] = firstNodeID).

Reply:
Yes that's true.. I wanted to test whether if we have more routes they all get deleted. I can create a 4th node to have more peers and do not add the first node. That way users will not be inclined to think this is acceptable behaviour.